File: python/0098_validate_binary_search_tree.py
# n = len(root)
# Time: O(n)
# Space: O(n)
from typing import Optional


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        if not root:
            return True
        inorder_list = []
        self.makeInorderList(root, inorder_list)
        for i in range(1, len(inorder_list)):
            if inorder_list[i] <= inorder_list[i-1]:
                return False
        return True

    def makeInorderList(self, root, inorder_list):
        if not root:
            return
        self.makeInorderList(root.left, inorder_list)
        inorder_list.append(root.val)
        self.makeInorderList(root.right, inorder_list)
        return

File: go/0098_validate_binary_search_tree.go
package main

import "math"

// n = number of nodes in the tree
// Time:  O(n)
// Space: O(n)
func isValidBST(root *TreeNode) bool {
	return dfsValidate(root, math.MinInt, math.MaxInt)
}

func dfsValidate(root *TreeNode, low int, high int) bool {
	if root == nil {
		return true
	}

	if low < root.Val && root.Val < high {
		return dfsValidate(root.Left, low, root.Val) && dfsValidate(root.Right, root.Val, high)
	} else {
		return false
	}
}
